service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/BlobCopier.java

```diff
@@ -0,0 +1,109 @@
+   * @param destinationStorageAccount Azure storage account for the destination container
+   * @param sourceContainer Azure storage container containing the blobs to be copied
+   * @param destinationContainer Azure storage container that will receive the copied blobs
+   * @return BlobCopyResult containing the status of each copy operation
+   */
+  public BlobCopierResult copyBlobs(
+      ControlledAzureStorageResource sourceStorageAccount,
+      ControlledAzureStorageResource destinationStorageAccount,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ControlledAzureStorageContainerResource destinationContainer) {
+    var sourceBlobContainerClient =
+        storageAccessService.buildBlobContainerClient(sourceContainer, sourceStorageAccount);
+    var destinationBlobCOntainerClient =
+        storageAccessService.buildBlobContainerClient(
+            destinationContainer, destinationStorageAccount);
+
+    // filter out any zero-length blobs, these are not copyable
+    var blobItems =
+        sourceBlobContainerClient.listBlobs().stream()
+            .filter(blobItem -> blobItem.getProperties().getContentLength() > 0);
+
+    var blobPollers =
+        blobItems.map(
+            sourceBlobItem ->
+                beginCopyBlob(
+                    storageAccessService,
+                    sourceContainer,
+                    sourceStorageAccount,
+                    userRequest,
+                    sourceBlobItem,
+                    sourceBlobContainerClient,
+                    destinationBlobCOntainerClient));
+
+    logger.info(
+        "Copying blobs [source_container_id = {}, source_workspace_id = {}, destination_container_id = {}, destination_workspace_id={}]",
+        sourceContainer.getResourceId(),
+        sourceContainer.getWorkspaceId(),
+        destinationContainer.getResourceId(),
+        destinationContainer.getWorkspaceId());
+    var pollResults = blobPollers.map(blobPoller -> blobPoller.waitForCompletion(MAX_POLL_TIMEOUT));
+    logger.info(
+        "Finished copying blobs [source_container_id = {}, source_workspace_id = {}, destination_container_id = {}, destination_workspace_id={}]",
+        sourceContainer.getResourceId(),
+        sourceContainer.getWorkspaceId(),
+        destinationContainer.getResourceId(),
+        destinationContainer.getWorkspaceId());
+
+    var resultsByStatus = pollResults.collect(groupingBy(PollResponse::getStatus));
+    return new BlobCopierResult(resultsByStatus);
+  }
+
+  private SyncPoller<BlobCopyInfo, Void> beginCopyBlob(
+      AzureStorageAccessService storageAccessService,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ControlledAzureStorageResource sourceStorageAccount,
+      AuthenticatedUserRequest userRequest,
+      BlobItem sourceBlobItem,
+      BlobContainerClient sourceBlobContainerClient,
+      BlobContainerClient destBlobContainerClient) {
+    var sasBundle =
```

Comment:
I think these types should be spelled out. Our rule of thumb is: only use `var` when the rhs tells you what the type is.

Reply:
Will do.